File: ZigZag_Conversion/sol2.cpp
/* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name : sol2.cpp
* Purpose : in-place solution
* Creation Date : 26-01-2014
* Last Modified : Sun Jan 26 12:21:15 2014
* Created By : wdd 
_._._._._._._._._._._._._._._._._._._._._.*/
#include "general.h"
    string convert(string s, int nRows) {
        if(nRows == 1) return s;
        string res = "";
        for(int i = 0; i < nRows; ++ i){ //search by row
            for(int j = 0; ; ++ j){
                int idx = i+2*(nRows-1)*j;//the string idx for the row is i+2*(nRows-1)*j
                int itv = idx - 2*i;//the string idx before is idx - 2*i
                if(i != 0 && i != nRows-1 && itv >= 0 && itv < s.length())
                    res += s[itv];
                if(idx < s.length())
                    res += s[idx];
                else
                    break;
            }
        }
        return res;
    }
int main(int argc, const char* argv[])
{
	return 0;
}

File: ZigZag_Conversion/sol2.py
#! /usr/bin/python
# --------------------------------------- 
# File Name : sol2.py
# Creation Date : 24-11-2014
# Last Modified : Mon Nov 24 16:04:15 2014
# Created By : wdd 
# --------------------------------------- 
import sys, re, string, pickle
from itertools import *
class Solution:
    # @return a string
    def convert(self, s, nRows):
        if nRows == 1: return s
        res = ""
        for i in range(0, nRows):
            j = 0;
            while 1:
                idx = i + j*2*(nRows-1)
                idx_b = idx - 2*i
                if idx_b > 0 and idx_b < len(s) and i and i != nRows-1:
                    res += s[idx_b]
                if idx < len(s):
                    res += s[idx]
                else: break
                j += 1
        return res
if __name__ == "__main__": main()
